From 08994c45b144982ccf51f2011d8a094fde4e36d3 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Thu, 7 Apr 2022 16:12:56 +0800
Subject: [PATCH 2147/2592] [Backport] memcg: Fix inconsistent oom event
 behavior for OOM_MEMCG_KILL

hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4X0YD

Reference: https://gitee.com/openeuler/kernel/commit/07c863f7cc45ecc1652a2813554b65cfe437363d

--------------------------------

Since memory.event is fully supported in cgroupv1, the problem of inconsistent
oom event behavior for OOM_MEMCG_KILL occurs again.
We fix the problem by add a new condition to support the event adding
continue. Therefore, there are two condition:
1) memcg is not root memcg;
2) the memcg is root memcg and the event is OOM_MEMCG_KILL of cgroupv1

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
Reviewed-by: Xiu Jianfeng <xiujianfeng@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
---
 include/linux/memcontrol.h | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 796cd8aad629..3a4ea9ed39fd 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -1198,6 +1198,18 @@ static inline void count_memcg_event_mm(struct mm_struct *mm,
 	rcu_read_unlock();
 }
 
+static bool memcg_event_add(struct mem_cgroup *memcg,
+			    enum memcg_memory_event event)
+{
+	if (!mem_cgroup_is_root(memcg))
+		return true;
+
+	if (event == MEMCG_OOM_KILL && !cgroup_subsys_on_dfl(memory_cgrp_subsys))
+		return true;
+
+	return false;
+}
+
 static inline void memcg_memory_event(struct mem_cgroup *memcg,
 				      enum memcg_memory_event event)
 {
@@ -1218,7 +1230,7 @@ static inline void memcg_memory_event(struct mem_cgroup *memcg,
 		if (cgrp_dfl_root.flags & CGRP_ROOT_MEMORY_LOCAL_EVENTS)
 			break;
 	} while ((memcg = parent_mem_cgroup(memcg)) &&
-		 !mem_cgroup_is_root(memcg));
+		 memcg_event_add(memcg, event));
 }
 
 static inline void memcg_memory_event_mm(struct mm_struct *mm,
-- 
2.17.1

